Python/specialize.c

```diff
@@ -633,3 +635,71 @@ _Py_Specialize_LoadGlobal(
     cache0->counter = saturating_start();
     return 0;
 }
+
+/* TODO:
+    - Specialize calling C types like int() with CALL_CTYPE
+    - Specialize python function calls.
+*/
+int
+_Py_Specialize_CallFunction(PyObject **stack_pointer, uint8_t original_oparg,
+    _Py_CODEUNIT *instr, SpecializedCacheEntry *cache)
+{
+    PyObject *callable = stack_pointer[-(original_oparg + 1)];
+    _PyAdaptiveEntry *cache0 = &cache->adaptive;
+    _PyCallFunctionCache *cache1 = &cache[-1].call_function;
+    /* Specialize C functions */
+    if (PyCFunction_CheckExact(callable)) {
+        PyCFunctionObject *meth = (PyCFunctionObject *)callable;
+        if (meth->m_ml == NULL) {
+            goto fail;
+        }
+        const char *name_ascii = meth->m_ml->ml_name;
+        _BuiltinCallKinds kind = -1;
+        switch (PyCFunction_GET_FLAGS(meth) & (METH_VARARGS | METH_FASTCALL |
+            METH_NOARGS | METH_O | METH_KEYWORDS | METH_METHOD)) {
+            case METH_VARARGS:
+                kind = PYCFUNCTION;
+                break;
+            case METH_VARARGS | METH_KEYWORDS:
+                kind = PYCFUNCTION_WITH_KEYWORDS;
+                break;
+            case METH_FASTCALL:
+                kind = _PYCFUNCTION_FAST;
+                break;
+            case METH_FASTCALL | METH_KEYWORDS:
+                kind = _PYCFUNCTION_FAST_WITH_KEYWORDS;
+                break;
+            case METH_NOARGS:
+                kind = PYCFUNCTION_NOARGS;
+                break;
+            case METH_O:
+                kind = PYCFUNCTION_O;
+                break;
+            /* This case should never happen with PyCFunctionObject -- only
+               PyMethodObject. See zlib.compressobj()'s methods for an example.
+            */
+            case METH_METHOD | METH_FASTCALL | METH_KEYWORDS:
+                // kind = PYCMETHOD;
+            default:
+                SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, "bad call flags");
+                goto fail;
+        }
+        assert(kind > 0);
+        PyCFunction cfunc = PyCFunction_GET_FUNCTION(meth);
+        assert(cfunc != NULL);
+        *instr = _Py_MAKECODEUNIT(CALL_CFUNCTION, _Py_OPARG(*instr));
+        cache0->index = (uint16_t)kind;
+        cache1->cfunc = cfunc;
+        goto success;
+    }
```

Comment:
Add a few `SPECIALIZATION_FAIL` here (guarded by `#if SPECIALIZATION_STATS`) so we get a view into what specializations would be profitable. I'd suggest classes, python functions, bound methods and method descriptors to start with.

Reply:
Stats for Python startup, the next most common cases are:
1. Python function calls.
2. Calls to`` __builtins__`` types, like ``map()``.
3. Bound method calls
4. python class inits
5. ``weakref.weakref``

Just the first 2 cases plus the current specializations seem to cover at least 80% of ``CALL_FUNCTION``. 3. is somewhat rare, and 4. exceedingly so. 5. only occurs once.

I'm going to try implement 1. and 2 and test. If those do not move pyperformance or some other realistic benchmark, then I suspect ``CALL_FUNCTION`` isn't a good candidate for specialization. So I'm marking this PR as draft till then.